syntax = "proto3";

import "google/protobuf/empty.proto";
package request;

// TODO: implement get_last_n
service KarlHost {
  rpc StartCompute (ComputeRequest) returns (NotifyStart);
  rpc Network (NetworkAccess) returns (NetworkAccessResult);
  rpc Get (GetData) returns (GetDataResult);
  rpc GetEvent (GetEventData) returns (GetDataResult);
  rpc Push (PushData) returns (google.protobuf.Empty);
}

service KarlController {
  // hosts
  rpc HostRegister (HostRegisterRequest) returns (HostRegisterResult);
  rpc ForwardNetwork (NetworkAccess) returns (google.protobuf.Empty);
  rpc ForwardGet (GetData) returns (GetDataResult);
  rpc ForwardPush (PushData) returns (google.protobuf.Empty);
  rpc ForwardState (StateChange) returns (google.protobuf.Empty);
  rpc FinishCompute (NotifyEnd) returns (google.protobuf.Empty);
  rpc Heartbeat (HostHeartbeat) returns (google.protobuf.Empty);

  // sensors
  rpc SensorRegister (SensorRegisterRequest) returns (SensorRegisterResult);
  rpc PushRawData (SensorPushData) returns (google.protobuf.Empty);
  rpc StateChanges (StateChangeInit) returns (stream StateChangePair);
}

/****************************************************************************
 * HostService
 ****************************************************************************/

message ComputeRequest {
  string host_token = 1;

  // caching
  string module_id = 2;
  bool cached = 3;
  bytes package = 4;

  // arguments
  string binary_path = 5;
  repeated string args = 6;
  repeated string envs = 7;

  // module inputs/outputs and network permissions
  string params = 8;
  string returns = 9;
  repeated string network_perm = 10;

  // automatically pass triggered data
  string triggered_tag = 11;
  string triggered_timestamp = 12;
  bytes triggered_data = 13;
}

message NotifyStart {
  string process_token = 1;
}

message KeyValuePair {
  bytes key = 1;
  bytes value = 2;
}

message NetworkAccess {
  string host_token = 1;
  string process_token = 2;
  string domain = 3;
  string method = 4;
  repeated KeyValuePair headers = 5;
  bytes body = 6;
}

message NetworkAccessResult {
  uint32 status_code = 1;
  repeated KeyValuePair headers = 2;
  bytes data = 3;
}

message GetData {
  string host_token = 1;
  string process_token = 2;
  string tag = 3;
  string lower = 4;
  string upper = 5;
}

message GetEventData {
  string process_token = 1;
  string tag = 2;
}

message GetDataResult {
  repeated string timestamps = 1;
  repeated bytes data = 2;
}

message PushData {
  string host_token = 1;
  string process_token = 2;
  string tag = 3;
  bytes data = 4;
}

message StateChangeInit {
  string sensor_token = 1;
}

message StateChangePair {
  string key = 1;
  bytes value = 2;
}

message StateChange {
  string host_token = 1;
  string process_token = 2;
  string sensor_id = 3;
  string key = 4;
  bytes value = 5;
}

/****************************************************************************
 * ControllerService
 ****************************************************************************/

// hosts

message HostRegisterRequest {
  string host_id = 1;
  string ip = 2;
  uint32 port = 3;
  string password = 4;
}

message HostRegisterResult {
  string host_token = 1;
}

message NotifyEnd {
  string host_token = 1;
  string process_token = 2;
}

message HostHeartbeat {
  string host_token = 1;
}

// sensors

message SensorRegisterRequest {
  string global_sensor_id = 1;
  bytes app = 2;
  repeated string returns = 3;
  repeated string keys = 4;
}

message SensorRegisterResult {
  string sensor_token = 1;
  string sensor_id = 2;
}

message SensorPushData {
  string sensor_token = 1;
  string param = 2;
  bytes data = 3;
}
